Examples/file-manager.lisp

```diff
@@ -0,0 +1,260 @@
+  (adopt directory (make-file name)))
+
+(define-presentation-to-command-translator directory->com-new-file
+    (directory* com-new-file file-manager
+                :gesture :select :priority -1
+                :documentation ((object stream) (format stream "New file")))
+    (directory)
+  `(,directory ,(accept 'string :prompt "name")))
+
+(define-command (com-new-directory :name t :command-table file-manager)
+    ((directory directory*)
+     (name      string))
+  (adopt directory (make-directory name)))
+
+(define-presentation-to-command-translator directory->com-new-directory
+    (directory* com-new-directory file-manager
+                :gesture :select :priority -1
+                :documentation ((object stream) (format stream "New directory")))
+    (directory)
+  `(,directory ,(accept 'string :prompt "name")))
+
+(defun drag-file-feedback (frame from-presentation stream x0 y0 x1 y1 state mode)
+  (case state
+    (:highlight
+     (with-output-recording-options (stream :record nil :draw t)
+       (multiple-value-bind (old-x old-y) (output-record-position from-presentation)
+         (let ((offset-x (- old-x x0))
+               (offset-y (- old-y y0))
+               (width    (bounding-rectangle-width from-presentation))
+               (height   (bounding-rectangle-height from-presentation)))
+                                        ; (erase-output-record from-presentation stream)
+           (setf (output-record-position from-presentation)
+                 (values (+ x1 offset-x) (+ y1 offset-y)))
+                                        ; (add-output-record from-presentation (stream-output-history stream))
+           (replay-output-record from-presentation stream)
+           (setf (output-record-position from-presentation) (values old-x old-y))
+           #+no (draw-rectangle* stream x1 y1 (+ x1 width) (+ y1 height)
+                                 :filled nil :line-dashes '(4 4))
+           (let ((text (string-downcase mode)))
+             (draw-text* stream text (+ x1 offset-x width 4) (+ y1 offset-y height 4)))))))
+    (:unhighlight
+     (multiple-value-bind (old-x old-y) (output-record-position from-presentation)
+       (let* ((offset-x (- old-x x0))
+              (offset-y (- old-y y0))
+              (width      (bounding-rectangle-width from-presentation))
+              (height     (bounding-rectangle-height from-presentation))
+              (text       (string-downcase mode))
+              (text-width (text-size stream text)))
+         (repaint-sheet stream (make-rectangle* (+ x1 offset-x -1)
+                                                (+ y1 offset-y -1)
+                                                (+ x1 offset-x width 2 4 text-width)
+                                                (+ y1 offset-y height 2 4))))))))
+
+(defun drag-file-feedback/copy (frame from-presentation stream x0 y0 x1 y1 state)
+  (drag-file-feedback frame from-presentation stream x0 y0 x1 y1 state :copy))
```

Comment:
when I copy directory I get a file with the same name (but not a directory)

Reply:
either that or symlink-directory-on-copy or refusing to copy a directory (with a notification, that it has files and copy is not recursive).

Reply:
recursive copy directory to one of its ancestors blows the stack